src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -337,93 +303,32 @@ class ShenandoahDirectCardMarkRememberedSet: public CHeapObj<mtGC> {
-//        entirety (because the thread that is responsible for
-//        scanning the object cannot rely upon the card-table entries
-//        associated with the following cluster).  Whenever references
-//        to young-gen memory are found within the scanned data, the
-//        associated overreach card table entries are marked as dirty
-//        by the scanning thread.
-//  3. If a cluster is spanned in its entirety by an object that
-//     originates within a preceding cluster's memory, the thread
-//     assigned to examine this cluster does absolutely nothing.  The
-//     thread assigned to scan the cluster that holds the object's
-//     starting address takes full responsibility for scanning the
-//     entire object and updating the associated card-table entries.
-//  4. If a cluster is spanned partially by an object that originates
-//     within a preceding cluster's memory, the thread assigned to
-//     examine this cluster marks the card-table entry as clean for
-//     each card table that is fully spanned by this overreaching
-//     object.  If a card-table entry's memory is partially spanned
-//     by the overreaching object, the thread sets the card-table
-//     entry to clean if it was previously dirty and if the portion
-//     of the card-table entry's memory that is not spanned by the
-//     overreaching object does not hold pointers to young-gen
-//     memory.
-//  5. While examining a particular card belonging to a particular
-//     cluster, if an object reaches beyond the end of its card
-//     memory, the thread "scans" all portions of the object that
-//     correspond to DIRTY card entries within the current cluster and
-//     all portions of the object that reach into following clustesr.
-//     After this object is scanned, continue scanning with the memory
-//     that follows this object if this memory pertains to the same
-//     cluster.  Otherwise, consider this cluster's memory to have
-//     been fully examined.
+//  1. The thread responsible for scanning the cards in a cluster modifies
+//     the associated card-table entries. Only cards that are dirty are
+//     processed, except as described below for the case of objects that
+//     straddle more than one card.
+//  2. Object Arrays are precisely dirtied, so only the portion of the obj-array
+//     that overlaps the range of dirty cards in its cluster are scanned
+//     by each worker thread. This holds for portions of obj-arrays that extend
+//     over clusters processed by different workers, with each worked responsible
+//     for scanning the portion of the obj-array overlapping the dirty cards in
+//     its cluster.
+//  3. Non-array objects are precisely dirtied by the interpreter and the compilers
+//     (why? Are offsets of a field in an object that expensive to determine?).
+//     For such objects that extend over multiple cards, or even multiple clusters,
```

Comment:
Historically, we borrowed the card-marking barrier from existing generational GC implementations and did not want to burden ourselves with trying to change it.  Presumably, experience with other GCs demonstrates that this works "well enough".  It would appear that non-array objects are usually not "extremely large".

Reply:
I realize that my comments in lines 324-328 were aspirational for changes I wanted to make but which is not the case today. I'll correct those and a few other such "aspirational" comments that got left behind in the code as I was working on subsequent changes.

It is true what you say that non-array objects are usually not very large (unless the result of code  generated by frameworks such as, e.g.,  protobufs).

Reply:
Fixed.